File: backend/app/services/consensus.py
"""Consensus computation: weighted-median dose with minority-protocol detection.

Mirrors the TypeScript implementation in packages/core/src/consensus.ts — keep
these in sync when either changes. This runs server-side as a nightly job to
produce ConsensusSnapshot rows that mobile reads.
"""
from dataclasses import dataclass

from app.models.community import DoseLogPost

# Weight brackets in kg — 10 kg wide, cover typical adult range.
WEIGHT_BRACKETS: list[tuple[float, float]] = [
    (50, 70),
    (60, 80),
    (70, 90),
    (80, 100),
    (90, 110),
    (100, 120),
    (110, 130),
]


@dataclass
class PostForConsensus:
    dose_mcg: float
    doses_per_week: float
    user_weight_kg: float
    bloodwork_attached: bool
    body_composition_attached: bool
    batch_info_attached: bool
    longitudinal: bool
    vendor_flagged: bool


def post_weight(p: PostForConsensus) -> float:
    if p.vendor_flagged:
        return 0.0
    w = 1.0
    if p.bloodwork_attached:
        w *= 5.0
    elif p.body_composition_attached:
        w *= 3.0
    if p.batch_info_attached:
        w *= 2.0
    if p.longitudinal:
        w *= 2.0
    return w


def weighted_quantile(values: list[float], weights: list[float], q: float) -> float:
    if not values:
        return 0.0
    pairs = sorted(
        [(v, w) for v, w in zip(values, weights) if w > 0],
        key=lambda x: x[0],
    )
    if not pairs:
        return 0.0
    total = sum(w for _, w in pairs)
    if total == 0:
        return pairs[len(pairs) // 2][0]
    target = total * q
    cum = 0.0
    for v, w in pairs:
        cum += w
        if cum >= target:
            return v
    return pairs[-1][0]


def compute_cluster_stats(
    posts: list[PostForConsensus],
) -> tuple[int, float, float, float, list[dict]]:
    """Return (n, median, p25, p75, minority_protocols) in mcg/kg/week."""
    per_kg_per_week = [
        (p.dose_mcg * p.doses_per_week) / p.user_weight_kg for p in posts
    ]
    weights = [post_weight(p) for p in posts]
    median = weighted_quantile(per_kg_per_week, weights, 0.5)
    p25 = weighted_quantile(per_kg_per_week, weights, 0.25)
    p75 = weighted_quantile(per_kg_per_week, weights, 0.75)

    total_weight = sum(weights)
    minority: list[dict] = []
    if total_weight > 0:
        for side in ('lower', 'higher'):
            subset = [
                (p, w, v)
                for p, w, v in zip(posts, weights, per_kg_per_week)
                if (v < median if side == 'lower' else v > median)
            ]
            sub_w = sum(w for _, w, _ in subset)
            if sub_w / total_weight < 0.15:
                continue
            sub_median = weighted_quantile(
                [v for _, _, v in subset],
                [w for _, w, _ in subset],
                0.5,
            )
            if abs(sub_median - median) / (median or 1) < 0.25:
                continue
            minority.append(
                {
                    'label': f'{side}-dose',
                    'median_mcg_per_kg_per_week': sub_median,
                    'weight_share': sub_w / total_weight,
                    'n': len(subset),
                }
            )

    return len(posts), median, p25, p75, minority


async def rebuild_snapshots_for_peptide(peptide_slug: str) -> int:
    """Rebuild all consensus snapshots for a peptide. Returns number written."""
    from app.models.community import ConsensusSnapshot

    all_posts = await DoseLogPost.filter(
        peptide_slug=peptide_slug,
        vendor_flagged=False,
        hidden=False,
    ).all()

    goals_in_data = sorted({p.goal for p in all_posts})
    written = 0

    for goal in goals_in_data:
        goal_posts = [p for p in all_posts if p.goal == goal]
        for w_min, w_max in WEIGHT_BRACKETS:
            bracket_posts = [p for p in goal_posts if w_min <= p.user_weight_kg <= w_max]
            if not bracket_posts:
                continue
            pc = [
                PostForConsensus(
                    dose_mcg=p.dose_mcg,
                    doses_per_week=p.doses_per_week,
                    user_weight_kg=p.user_weight_kg,
                    bloodwork_attached=p.bloodwork_attached,
                    body_composition_attached=p.body_composition_attached,
                    batch_info_attached=p.batch_info_attached,
                    longitudinal=p.longitudinal,
                    vendor_flagged=p.vendor_flagged,
                )
                for p in bracket_posts
            ]
            n, median, p25, p75, minority = compute_cluster_stats(pc)
            await ConsensusSnapshot.update_or_create(
                peptide_slug=peptide_slug,
                goal=goal,
                weight_bracket_kg_min=w_min,
                defaults={
                    'weight_bracket_kg_max': w_max,
                    'n_posts': n,
                    'median_mcg_per_kg_per_week': median,
                    'p25_mcg_per_kg_per_week': p25,
                    'p75_mcg_per_kg_per_week': p75,
                    'minority_protocols': minority,
                    'low_confidence': n < 5,
                },
            )
            written += 1

    return written


async def rebuild_all_snapshots() -> int:
    """Nightly job: rebuild snapshots across every peptide with posts."""
    slugs = await DoseLogPost.all().distinct().values_list('peptide_slug', flat=True)
    total = 0
    for slug in slugs:
        total += await rebuild_snapshots_for_peptide(slug)
    return total

File: packages/core/src/consensus.ts
/**
 * Community consensus math.
 *
 * Users post structured dose logs with outcomes. The consensus engine
 * aggregates them using post-level trust weights (bloodwork attached → ×5,
 * anonymous text only → ×1) and returns a weighted median with IQR bands,
 * plus minority-protocol detection when a significant cluster diverges.
 *
 * Philosophy: surface dissent, don't hide it. If 20% of users cluster at a
 * lower dose with better outcomes, that's the "minority protocol" and it
 * deserves its own row on the UI.
 */

export interface PostEvidence {
  /** Bloodwork attached — labs covering the relevant biomarker panel. */
  bloodwork: boolean;
  /** Body composition measurement attached (DEXA, BodPod, calipers). */
  bodyComposition: boolean;
  /** Batch or COA info attached. */
  batchInfo: boolean;
  /** Post covers 60+ days of consistent logging. */
  longitudinal: boolean;
  /** Optional identity verification level. */
  verified: boolean;
  /** Post flagged as affiliate or vendor content. */
  vendorFlagged: boolean;
}

export interface CommunityPost {
  id: string;
  userId: string;
  peptideSlug: string;
  /** Dose per injection in mcg. */
  doseMcg: number;
  /** Doses per week (for weekly-exposure math). */
  dosesPerWeek: number;
  /** Cycle length in weeks (null = ongoing/unspecified). */
  weeksOn?: number | null;
  /** User's weight in kg at the time of the protocol. */
  weightKg: number;
  /** Goal category — e.g. 'injury_recovery'. */
  goal: string;
  /** Self-reported outcome score: -2 = worse, 0 = no change, 2 = strong improvement. */
  outcomeScore?: number;
  /** Reported side effect severity 0–10. */
  sideEffectSeverity?: number;
  evidence: PostEvidence;
}

export interface ConsensusInput {
  posts: CommunityPost[];
  /** Viewer's weight bracket for filtering — ±20% is applied. */
  viewerWeightKg?: number;
  /** Viewer's goal for filtering. */
  goal?: string;
}

export interface ConsensusResult {
  /** Number of posts used after filtering. */
  n: number;
  /** Weighted median dose in mcg/kg/week (for cross-weight comparability). */
  medianMcgPerKgPerWeek: number;
  /** P25 and P75 of the weighted distribution. */
  p25McgPerKgPerWeek: number;
  p75McgPerKgPerWeek: number;
  /** Scaled median dose for the viewer's weight. */
  scaledDoseMcg?: number;
  /** Minority protocols — secondary clusters with ≥15% weight share. */
  minorityProtocols: Array<{
    /** Cluster label: 'lower-dose' | 'higher-dose'. */
    label: string;
    medianMcgPerKgPerWeek: number;
    weightShare: number;
    n: number;
  }>;
  /** Warning if n is too small for reliable consensus. */
  lowConfidence: boolean;
}

export function postWeight(post: CommunityPost): number {
  if (post.evidence.vendorFlagged) return 0;
  let w = 1;
  if (post.evidence.bloodwork) w *= 5;
  else if (post.evidence.bodyComposition) w *= 3;
  if (post.evidence.batchInfo) w *= 2;
  if (post.evidence.longitudinal) w *= 2;
  if (post.evidence.verified) w *= 2;
  return w;
}

/**
 * Weighted median — linear interpolation between sorted weighted points.
 */
export function weightedQuantile(
  values: number[],
  weights: number[],
  q: number,
): number {
  if (values.length === 0) return 0;
  const sorted = values
    .map((v, i) => ({ v, w: weights[i] ?? 0 }))
    .filter((p) => p.w > 0)
    .sort((a, b) => a.v - b.v);
  if (sorted.length === 0) return 0;
  const total = sorted.reduce((s, p) => s + p.w, 0);
  if (total === 0) return sorted[Math.floor(sorted.length / 2)]!.v;
  const target = total * q;
  let cum = 0;
  for (let i = 0; i < sorted.length; i++) {
    cum += sorted[i]!.w;
    if (cum >= target) return sorted[i]!.v;
  }
  return sorted[sorted.length - 1]!.v;
}

/**
 * Compute weighted-median consensus from a set of community posts.
 *
 * Filtering:
 *   - Drops vendor-flagged posts (weight 0)
 *   - If viewerWeightKg given, filters to posts within ±20% weight
 *   - If goal given, filters to matching goal
 *
 * Returns dose normalized to mcg/kg/week for cross-weight comparability,
 * plus a scaled dose for the viewer.
 */
export function computeConsensus(input: ConsensusInput): ConsensusResult {
  let relevant = input.posts.filter((p) => !p.evidence.vendorFlagged);
  if (input.goal) relevant = relevant.filter((p) => p.goal === input.goal);
  if (input.viewerWeightKg != null) {
    const lo = input.viewerWeightKg * 0.8;
    const hi = input.viewerWeightKg * 1.2;
    relevant = relevant.filter((p) => p.weightKg >= lo && p.weightKg <= hi);
  }

  const perKgPerWeek = relevant.map(
    (p) => (p.doseMcg * p.dosesPerWeek) / p.weightKg,
  );
  const weights = relevant.map(postWeight);

  const median = weightedQuantile(perKgPerWeek, weights, 0.5);
  const p25 = weightedQuantile(perKgPerWeek, weights, 0.25);
  const p75 = weightedQuantile(perKgPerWeek, weights, 0.75);

  // Detect minority clusters: split posts into above-median and below-median,
  // report if either side has ≥15% of total weight AND its own median is
  // ≥25% different from the overall.
  const totalWeight = weights.reduce((s, w) => s + w, 0);
  const minorityProtocols: ConsensusResult['minorityProtocols'] = [];

  if (totalWeight > 0) {
    for (const side of ['lower', 'higher'] as const) {
      const subset = relevant
        .map((p, i) => ({ p, w: weights[i] ?? 0, v: perKgPerWeek[i] ?? 0 }))
        .filter((r) => (side === 'lower' ? r.v < median : r.v > median));
      const subW = subset.reduce((s, r) => s + r.w, 0);
      if (subW / totalWeight < 0.15) continue;
      const subMed = weightedQuantile(
        subset.map((r) => r.v),
        subset.map((r) => r.w),
        0.5,
      );
      if (Math.abs(subMed - median) / (median || 1) < 0.25) continue;
      minorityProtocols.push({
        label: side === 'lower' ? 'lower-dose' : 'higher-dose',
        medianMcgPerKgPerWeek: subMed,
        weightShare: subW / totalWeight,
        n: subset.length,
      });
    }
  }

  const scaledDoseMcg =
    input.viewerWeightKg != null
      ? (median * input.viewerWeightKg) / Math.max(1, relevant[0]?.dosesPerWeek ?? 1)
      : undefined;

  return {
    n: relevant.length,
    medianMcgPerKgPerWeek: median,
    p25McgPerKgPerWeek: p25,
    p75McgPerKgPerWeek: p75,
    scaledDoseMcg,
    minorityProtocols,
    lowConfidence: relevant.length < 5,
  };
}
